gmm.py

```diff
@@ -331,3 +411,92 @@ def __update_pi(self, pi):
         assert pi.size() in [(1, self.n_components, 1)], "Input pi does not have required tensor dimensions (%i, %i, %i)" % (1, self.n_components, 1)
 
         self.pi.data = pi
+
+        
+    @staticmethod
+    def get_kmeans_mu(X, n_centers, min_delta=1e-3, init_times=50):
+        """
+        input:
+        x:              torch.Tensor (n, d),(n, 1, d)
+        min_delta:      int, thresh of kmeans stop
+        init_times:     times try's to find better center
+        output:
+        center:         torch.Tensor (1, k, d)
+        """
+        if len(X.size()) == 3:
+            X = X.squeeze(1)
+        min_value, max_value = X.min(), X.max()
+        X = (X - min_value) / (max_value - min_value)
+        min_cost = np.inf
+        for i in range(init_times):
+            tmp_center = X[np.random.choice(np.arange(X.shape[0]), size=n_centers, replace=False), ...]
+            l2_dis = torch.norm((X.unsqueeze(1).repeat(1, n_centers, 1) - tmp_center), p=2, dim=2)
+            l2_cls = torch.argmin(l2_dis, dim=1)
+            cost = 0
+            for c in range(n_centers):
+                cost += torch.norm(X[l2_cls==c] - tmp_center[c],p=2,dim=1).mean()
+            if cost < min_cost:
+                min_cost = cost
+                center = tmp_center
+
+        delta = np.inf
+        while delta > min_delta:
+            l2_dis = torch.norm((X.unsqueeze(1).repeat(1,n_centers,1)-center),p=2,dim=2)
+            l2_cls = torch.argmin(l2_dis, dim=1)
+            center_old = center.clone()
+            for c in range(n_centers):
+                center[c] = X[l2_cls==c].mean(dim=0)
+            delta = torch.norm((center_old-center),dim=1).max()
+
+        return (center.unsqueeze(0)*(max_value-min_value)+min_value)
+
+if __name__=="__main__":
+    from math import sqrt
```

Comment:
This has now been moved to tests, correct?

Reply:
Maybe I should delete it.